master/internal/scheduler/default_resource_provider.go

```diff
@@ -80,92 +58,73 @@ func NewDefaultRP(
 }
 
-// assignTask allocates cluster data structures and sends the appropriate actor
-// messages to start a task if there are enough resources in the cluster to run
-// the task. If there are not, assignTask returns false.
-func (d *DefaultRP) assignTask(task *Task) bool {
-	fits := findFits(task, d.agents, d.fittingMethod)
+// allocateResources assigns resources based on a request and notifies the request
+// handler of the assignment. It returns true if it is successfully allocated.
+func (d *DefaultRP) allocateResources(req *AllocateRequest) bool {
+	fits := findFits(req, d.agents, d.fittingMethod)
 
 	if len(fits) == 0 {
 		return false
 	}
 
-	assignments := make([]Assignment, 0, len(fits))
-
+	allocations := make([]Allocation, 0, len(fits))
 	for _, fit := range fits {
-		assignments = append(
-			assignments,
-			d.assignContainer(task, fit.Agent, fit.Slots, len(fits)),
-		)
+		container := newContainer(req, fit.Agent, fit.Slots, len(allocations))
+		allocations = append(allocations, &containerAllocation{
+			req:       req,
+			agent:     fit.Agent,
+			container: container,
+			devices:   fit.Agent.allocateFreeDevices(fit.Slots, cproto.ID(container.id)),
+		})
 	}
 
-	task.handler.System().Tell(task.handler, TaskAssigned{
-		Assignments: assignments,
-	})
+	allocated := ResourcesAllocated{ID: req.ID, Allocations: allocations}
+	d.taskList.SetAllocations(req.TaskActor, &allocated)
+	req.TaskActor.System().Tell(req.TaskActor, allocated)
+	log.Infof("allocated resources to %s", req.TaskActor.Address())
 
 	return true
 }
 
-// terminateTask sends the appropriate actor messages to terminate a task and
-// deallocate its cluster data structures. The task may not be terminated if it
-// is in the right state unless forcible is true.
-func (d *DefaultRP) terminateTask(task *Task, forcible bool) {
-	switch {
-	case task.state == taskTerminated:
-		// The task has already been terminated so this is a noop.
-
-	case len(task.containers) == 0 || task.state == taskPending:
-		// The task is not running so there is no need to request the task to terminate. The task is
-		// marked as aborted.
-		d.taskTerminated(task, true)
-
-	case forcible:
-		// Notify the agent to kill the task.
-		task.mustTransition(taskTerminating)
-		for _, c := range task.containers {
-			if c.state != containerTerminated {
-				c.mustTransition(containerTerminating)
-			}
-			c.agent.handler.System().Tell(c.agent.handler, aproto.SignalContainer{
-				ContainerID: cproto.ID(c.id), Signal: syscall.SIGKILL})
-		}
+func (d *DefaultRP) releaseResource(handler *actor.Ref) {
+	// The request handler is removed so that it would not take up resources.
+	// In practice, the request handler might gracefully wait for the containers
+	// to exit themselves, which might take very long time or might not happen at all.
+	// In the mean time, resources are re-allocated to other requests, which might let
+	// the old container exiting and the new container run slower than normal.
```

Comment:
not really clear about what it means that new containers will run slower, does it mean they will get scheduled on the same slot as a task that not yet exited?

Reply:
This is not accurate anymore because I fixed when containers are killed a slot might be allocated to two containers.

Reply:
> not really clear about what it means that new containers will run slower, does it mean they will get scheduled on the same slot as a task that not yet exited?

yes. I removed the comment and the call to `RemoveTaskByHandler`